### PR TITLE
style: interaction data

### PR DESCRIPTION
**Pull request by beachasaurus-rex**
# Description

Changing the interaction data types so that they're more aligned with discord's API. The main goal of this refactoring is to help disgord users express what is actually happening during interactions. In other words, if the user intends to receive an [interaction object message component](https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object-message-component-data-structure), then the user should receive the equivalent from disgord.

[Interaction data](https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object-interaction-data) is polymorphic, so this PR shows 1 way to achieve that a bit better in disgord.

[Interaction callback data]() is also polymorphic, so this PR shows 1 way to achieve that a bit better in disgord.

## Breaking Change?

yes

## Benchmarks

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] Commented complex situations or referenced the discord documentation
- [ ] Updated documentation
- [ ] Added/Updated unit tests
- [ ] Added/Updated benchmarks (if this is a performance critical component)

### CODE REVIEW

**Comment by beachasaurus-rex** [comment]
I know that a good chunk of the internal stuff would need to change for this to even work, but I want to get feedback on at least the surface-level API changes I've made so far.

File: interaction.go [hunk 1]
```diff
@@ -47,6 +48,13 @@ const (
 	InteractionCallbackUpdateMessage
 )
 
+type Interactable interface {
```
Comment (andersfylling) [inline interaction.go:51]: Do we need this? Can't we just specify the exact parameter type instead?

File: client.go [hunk 2]
```diff
@@ -524,25 +525,35 @@ func (ih *internalHandlers) saveApplicationID(_ Session, evt *Ready) {
...
 //////////////////////////////////////////////////////
 
-func (c *Client) EditInteractionResponse(ctx context.Context, interaction *InteractionCreate, message *UpdateMessage) error {
-	postBody, contentType, err := message.prepare()
-	if err != nil {
-		return err
-	}
+func (c *Client) EditInteractionResponse(ctx context.Context, interaction Interactable, message *UpdateMessage) error {
+	switch interaction.(type) {
+	case *InteractionCreate[MessageComponentInteractionData]:
```
Comment (andersfylling) [inline client.go:530]: Looking at it, I'd say this is a nice use of generics, and I'm up for it. I assume we have to do something similar for the response

File: interaction.go (hunk 1 above)
Comment (beachasaurus-rex) [reply interaction.go:51]: `Session.SendInteraction()` can't be a generic function in the interface, so an interface needs to be passed, unless you prefer multiple iterations of `Session.SendInteraction()` for every type parameter of `InteractionData[T]`.

Comment (andersfylling) [reply interaction.go:51]: ah fair

**Pull request closed without merging**